crates/common/tedge_config_macros/src/all_or_nothing.rs

```diff
@@ -0,0 +1,87 @@
+use crate::OptionalConfig;
+
+pub fn all_or_nothing<T, U>(
+    t: OptionalConfig<T>,
+    u: OptionalConfig<U>,
+) -> Result<Option<(T, U)>, String> {
```

Comment:
Definitely simpler and better to have this defined as a regular function instead of a macro.

Beyond this specific case, I see this simplification as a better separation of the responsibilities. Adding constraints on what makes sense should be done at the application level while the config should be focus on a more syntactic level (what is set, what is missing, what is ill-formed).